// @directus/types/dist/websockets/collab.js
import {} from 'crypto';
import z from 'zod';
import { TYPE } from './type.js';
export const COLLAB_BUS = 'collab';
export const COLORS = ['purple', 'pink', 'blue', 'green', 'yellow', 'orange', 'red'];
export const ACTION = {
    CLIENT: {
        JOIN: 'join',
        LEAVE: 'leave',
        UPDATE: 'update',
        UPDATE_ALL: 'updateAll',
        FOCUS: 'focus',
        DISCARD: 'discard',
    },
    SERVER: {
        INIT: 'init',
        JOIN: 'join',
        LEAVE: 'leave',
        SAVE: 'save',
        DELETE: 'delete',
        UPDATE: 'update',
        FOCUS: 'focus',
        DISCARD: 'discard',
        ERROR: 'error',
    },
};
const BaseClientMessage = z.object({
    type: z.literal(TYPE.COLLAB),
    room: z.string(),
});
// Messages from client to server
export const ClientMessage = z.discriminatedUnion('action', [
    z.object({
        type: z.literal(TYPE.COLLAB),
        action: z.literal(ACTION.CLIENT.JOIN),
        collection: z.string(),
        item: z.union([z.string(), z.number()]).nullable(),
        version: z.string().nullable(),
        color: z.enum(COLORS).nullable().optional(),
        initialChanges: z.record(z.string(), z.any()).optional(),
    }),
    BaseClientMessage.extend({
        action: z.literal(ACTION.CLIENT.LEAVE),
    }),
    BaseClientMessage.extend({
        action: z.literal(ACTION.CLIENT.UPDATE),
        field: z.string(),
        changes: z.unknown().optional(),
    }),
    BaseClientMessage.extend({
        action: z.literal(ACTION.CLIENT.UPDATE_ALL),
        changes: z.record(z.string(), z.any()).optional(),
    }),
    BaseClientMessage.extend({
        action: z.literal(ACTION.CLIENT.FOCUS),
        field: z.string().nullable(),
    }),
    BaseClientMessage.extend({
        action: z.literal(ACTION.CLIENT.DISCARD),
    }),
]);
